python/xorbits/numpy/numpy_adapters/core.py

```diff
@@ -0,0 +1,77 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import warnings
+from typing import Any, Callable, Dict, Type
+
+import numpy as np
+
+from ...core.adapter import MarsOutputType, wrap_mars_callable
+from ...core.utils.fallback import wrap_fallback_module_method
+
+_NO_ANNOTATION_FUNCS: Dict[Callable, MarsOutputType] = {
+    np.random.default_rng: MarsOutputType.object,
+    np.random.Generator: MarsOutputType.object,
+    np.random.PCG64: MarsOutputType.object,
+    np.random.MT19937: MarsOutputType.object,
+    np.random.Generator: MarsOutputType.object,
+}
+
+
+def _get_output_type(func: Callable) -> MarsOutputType:
+    try:  # pragma: no cover
+        return_annotation = inspect.signature(func).return_annotation
+        if return_annotation is inspect.Signature.empty:
+            # mostly for python3.7 whose return_annotation is always empty
+            return _NO_ANNOTATION_FUNCS.get(func, MarsOutputType.object)
+        all_types = [t.strip() for t in return_annotation.split("|")]
+
+        return (
+            MarsOutputType.tensor if "ndarray" in all_types else MarsOutputType.object
+        )
+    except (
+        ValueError
+    ):  # some np methods return objects and inspect.signature throws a ValueError
+        return _NO_ANNOTATION_FUNCS.get(func, MarsOutputType.object)
+
+
+def _collect_numpy_module_members(np_cls: Type) -> Dict[str, Any]:
+    from ..mars_adapters.core import MARS_TENSOR_CALLABLES
+
+    module_methods: Dict[str, Any] = dict()
+    with warnings.catch_warnings():
+        for name, cls_member in inspect.getmembers(np_cls):
+            if (
+                name not in MARS_TENSOR_CALLABLES
+                and callable(cls_member)
+                and not name.startswith("_")
+            ):
+                warning_str = f"xorbits.numpy.{name} will fallback to NumPy"
+                output_type = _get_output_type(getattr(np_cls, name))
+
+                module_methods[name] = wrap_mars_callable(
+                    wrap_fallback_module_method(np_cls, name, output_type, warning_str),
+                    attach_docstring=True,
+                    is_cls_member=False,
+                    docstring_src_module=np_cls,
+                    docstring_src=getattr(np_cls, name, None),
+                    fallback_warning=True,
+                )
+
+    return module_methods
+
+
+NUMPY_MODULE_METHODS = _collect_numpy_module_members
```

Comment:
Would it be better just use the name `_collect_numpy_module_members`?

Intuitively, `NUMPY_MODULE_METHODS` should be something like `MARS_TENSOR_RANDOM_CALLABLES`, which is a dict that has been populated.

Reply:
yeah definitely. i will update the code

Reply:
since this is used in ```numpy/random/__init__.py```, i will rename the function to be ```collect_numpy_module_members``` and use it there